x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ShardsAllocationResults.java

```diff
@@ -9,6 +9,12 @@
 
 import org.elasticsearch.index.shard.ShardId;
 
+import java.util.List;
 import java.util.SortedSet;
 
-record ShardsSize(long sizeInBytes, SortedSet<ShardId> shardIds) {}
+record ShardsAllocationResults(
+    long sizeInBytes,
+    SortedSet<ShardId> shardIds,
+    List<NodeDecision> canAllocateDecisions,
+    List<NodeDecision> canRemainDecisions
```

Comment:
I wonder if we need to make this a `Map<ShardId, record<List<NodeDecision>, List<NodeDecision>>>` instead. That would make it possible to limit the number of shards to list to some count and to limit the number of decisions to list to a subset of those shards.

Reply:
I've created a new class `NodeDecisions` that holds `canAllocateDecisions` and `canRemainDecisions`. `ShardsAllocationResults` now holds these decisions by shard id as `Map<ShardId, NodeDecisions> shardNodeDecisions`.
This map gets stored as `unassignedNodeDecisions` and `assignedNodeDecisions` and then gets limited in the JSON output